docs/proposals/workflow.md

```diff
@@ -0,0 +1,255 @@
+}
+```
+
+
+#### `WorkflowSpec`
+
+```go
+// WorkflowSpec contains Workflow specification
+type WorkflowSpec struct {
+
+    // Optional duration in seconds the workflow needs to terminate gracefully. May be decreased in delete request.
+	// Value must be non-negative integer. The value zero indicates delete immediately.
+	// If this value is nil, the default grace period will be used instead.
+	// Set this value longer than the expected cleanup time for your workflow.
+    // If downstream resources (job, pod, etc.) define their TerminationGracePeriodSeconds
+    // the biggest is taken.
+	TerminationGracePeriodSeconds *int64 `json:"terminationGracePeriodSeconds,omitempty"`
+
+	// Steps contains the vertices of the workflow graph.
+	Steps []WorkflowStep `json:"steps,omitempty"`
+}
+```
+
+* `spec.steps`: is an array of `WorkflowStep`s.
+* `spec.terminationGracePeriodSeconds`: is the terminationGracePeriodSeconds.
+
+### `WorkflowStep`<sup>1</sup>
+
+The `WorkflowStep` resource acts as a [union](https://en.wikipedia.org/wiki/Tagged_union) of `JobSpec` and `ObjectReference`.
+
+```go
+// WorkflowStep contains necessary information to identifiy the node of the workflow graph
+type WorkflowStep struct {
+    // Id is the identifier of the current step
+    Id string  `json:"id,omitempty"`
+
+    // Spec contains the job specificaton that should be run in this Workflow.
+	// Only one between External and Spec can be set.
+	Spec JobSpec `json:"jobSpec,omitempty"`
+
+    // Dependecies represent dependecies of the current workflow step
+    Dependencies ObjectDependencies `json:"dependencies,omitempty"`
+
+	// External contains a reference to another schedulable resource.
+	// Only one between ExternalRef and Spec can be set.
+	ExternalRef api.ObjectReference `json:"externalRef,omitempty"`
+}
+```
+
+* `workflowStep.id` is a string to identify the current `Workflow`. The `workfowStep.id` is injected
+as a label in `metadata.annotations` in the `Job` created in the current step.
+* `workflowStep.jobSpec` contains the specification of the job to be executed.
+* `workflowStep.externalRef` contains a reference to external resources (for example another `Workflow`).
+
+```go
+type ObjectDependencies struct {
+    ...
+    ExistenceDependencies []ObjectReference `json:"existenceDependencies,omitempty"`
+    ControllerRef *ObjectReference `json:"controllerRef,omitempty"`
+    ...
+}
+```
+
+* `dependencies.controllerRef`: will contain the policy to trigger current `WorkflowStep`
+
+For `Workflow` basic scenario 2 controllers should be implemented: `AllDependenciesRunToCompletion`,
+`AtLeastOneDependencyRunToCompletion`. This approach permits implementation of other kinds of triggering
+policies, like for example data availability or other external event.
+
+
+
+### `WorkflowStatus`
+
+```go
+// WorkflowStatus contains the current status of the Workflow
+type WorkflowStatus struct {
+	Statuses []WorkflowStepStatus `json:statuses`
```

Comment:
Yes, we do. My understanding (which comes from our internal use of workflow) is that workflow is a pretty complex sequence of tasks. A user should have a way to modify it, even reassigning it (feature quite common in commercial tools) [see](https://github.com/kubernetes/kubernetes/pull/18827/files#diff-bda92e294a9d99318221126e00f42f58R243). Since the creation is quite a heavy process.